### ISSUE TITLE
How to start?

### ISSUE THREAD

**Issue by suve** [PascalGameDevelopment/SDL3-for-Pascal#1]
SDL3 is near release, with the [3.1.3 preview version](https://github.com/libsdl-org/SDL/releases/tag/preview-3.1.3) having been released earlier this week. As such, we can start getting ready to work on the SDL3 units.

So, my question is - how should we start? I see three options here:

1) Base our SDL3 units on the SDL2 units we already have. Use the [SDL3 migration guide](https://wiki.libsdl.org/SDL3/README/migration) to help us identify which parts need to be changed.

2) Start anew. Use some tool like [h2pas](https://www.freepascal.org/tools/h2pas.var) to automatically convert the C headers into Pascal units and then tweak the output until we arrive at something we're satisfied with.

3) Start from scratch, manually.

Option 1) would definitely be the easiest, though I am rather worried about the risk of us missing things here and there, leading to hard-to-find bugs where the units compile fine, but the definitions don't match what's in the C headers. As such, I'd be willing to devote some time to try out option 2) and see how that works out.

**Comment by Free-Pascal-meets-SDL-Website** [comment]
I totally agree with everything you said. If there was plenty of time I would prefer opt. 2, too. - But also I'd like to mention that just updating the sdl2 repo to version 2.0.14 for every unit took around 3 years. My experience with h2pas is that is still needs a lot of manual attention; I didn't find better tools though.  

For personal reasons I had and have few spare time in the last and coming months. Hence, for opt. 2 probably you would have to do a larger part. If you are willing to do so, I will contribute as much as possible, too (as always :D). 

The big benefits of opt. 2 are: We could still rely on many "rules" we established for the sdl2 repo. We could freely decide to change some bothering things (see below). And we would know about "every corner" of the code. 

If we go for opt. 2:
The first steps would be to decide for 

- general folder/file/code structure
- a code style (have a guideline)
- a license 

To my mind, actually we could heavily replicate the folder structure, file structure and code style from the sdl2 repo. We decided for these in many discussions already.

Some things to consider, that I would like to change compared to the sdl2 repo:
- decide for uniform file headers for every file (similar to sdl3's file headers)
- have the inc-file names being identical to the h-file names ("SDL_events.inc" instead of "sdlevents.inc")
- for the license I would go JUST for the zlib license (like the sdl3 project itself)

What are your thoughts on this?

Best regards
Matthias

**Comment by suve** [comment]
> My experience with h2pas is that is still needs a lot of manual attention; I didn't find better tools though.

Well, yeah, I don't really expect it to produce something that'll be 100% production-ready; but anything that reduces the amount of manual work is a good idea. 

> decide for uniform file headers for every file

I'm not sure what do you mean by this. Uniform licence info headers?

> have the inc-file names being identical to the h-file names ("SDL_events.inc" instead of "sdlevents.inc")

Minor change, but yeah, doesn't affect anything from the user's perspective and makes the repo more consistent with the original sources.

> for the license I would go JUST for the zlib license (like the sdl3 project itself)

zlib licence is fine by me.

**Comment by Free-Pascal-meets-SDL-Website** [comment]
> Well, yeah, I don't really expect it to produce something that'll be 100% production-ready; but anything that reduces the amount of manual work is a good idea.

Absolutely!

> I'm not sure what do you mean by this. Uniform licence info headers?

Right, also project name, project url. 

My suggestion:

```
SDL3 for Pascal
Copyright (c) 2024 Pascal Game Development Team (https://github.com/PascalGameDevelopment/SDL3-for-Pascal)

This software is provided 'as-is', without any express or implied warranty. In no event will the authors be held liable for any damages arising from the use of this software.

Permission is granted to anyone to use this software for any purpose, including commercial applications, and to alter it and redistribute it freely, subject to the following restrictions:

1. The origin of this software must not be misrepresented; you must not claim that you wrote the original software. If you use this software in a product, an acknowledgment in the product documentation would be appreciated but is not required.
2. Altered source versions must be plainly marked as such, and must not be misrepresented as being the original software.
3. This notice may not be removed or altered from any source distribution.
```

Another question, which arises: How do we do testing? For the sdl2 repo I suggested a testing framework (by the PR). We could also go for a simpler approach as we do now. What are your thoughts on this topic?

**Comment by suve** [comment]
On one hand, I'm okay with the header. On the other hand, I think I'd prefer keeping the licence in a `LICENSE` file and just having the header point people to said file. Or we could just slap an [SPDX license tag](https://spdx.org/licenses/) on it:
```
SPDX-License-Identifier: Zlib
```

> How do we do testing? For the sdl2 repo I suggested a testing framework (by the PR)

Uh, yeah, sorry for not reviewing that. :sweat_smile:  I know that FPC ships with some testing framework, but it's largely undocumented; I'd prefer to look into this tool first, instead of dragging such a large 3-rd party dependency into the repo.

**Comment by Free-Pascal-meets-SDL-Website** [comment]
So, what about

```
SDL3 for Pascal
(https://github.com/PascalGameDevelopment/SDL3-for-Pascal)
SPDX-License-Identifier: Zlib

```
?

> Uh, yeah, sorry for not reviewing that. 😅 I know that FPC ships with some testing framework, but it's largely undocumented; I'd prefer to look into this tool first, instead of dragging such a large 3-rd party dependency into the repo.

No harm done. - Well, according to my research this 3rd party repo should be the better choice compared to FP's own testing framework. That said, I also don't like the fact that the 3rd-party repo is so large though. I will reconsider this; maybe we go with the FP framework for convenience or use the simple tests like before. For the moment we should focus on other tasks anyway. :)

I will craft the initial commit soon, so we can start. :D

**Comment by sechshelme** [comment]
I just saw that you also want to get to grips with SDL3. I'm already working on it and it's almost finished, the header already runs with SDL 3.1.6
I did it from scratch.

https://github.com/sechshelme/Lazarus-SDL3.0-Packages_and_Examples

**Comment by Free-Pascal-meets-SDL-Website** [comment]
@sechshelme Thanks for letting us know. Impressive work! Well done! Maybe we could collab in a way, any suggestions?

@suve How should we proceed now? I'm kind of undecided right now knowing that there is repo having done a lot of effort already. What are your thoughts?

**Comment by sechshelme** [comment]
>@sechshelme Thanks for letting us know. Impressive work! Well done! Maybe we could collab in a way, any suggestions?

My binding should be fully operational for SDL 3.1.6.

It's just the SDL3_xxx stuff that's causing problems, as things in the c-libs have been changed in the meantime.

As soon as these appear in the releases, I'll recompile them.

**Comment by sechshelme** [comment]
I just compiled the new SD3_images 3.1.0.
I suspect it doesn't work anymore, and the other sdl3_xxx will come.

**Comment by Free-Pascal-meets-SDL-Website** [comment]
Thanks for letting us know. Impressive work!

@suve Please check your mails.

**Comment by Free-Pascal-meets-SDL-Website** [comment]
@sechshelme Although I really appreciate your impressive work on the sdl3 bindings, we decided to start a community driven project with at times different characteristics. We whole-heartedly invite you to join us here though :), that would be really great! 

Best regards
Matthias
